### DISCUSSION TITLE
Only 100 results?

### DISCUSSION THREAD

**Discussion by lsegal** [community/community#9868] [Other Feature Feedback, Questions, & Ideas] [answered]
I am noticing that most searches are only yielding 5 pages of results. Is this just a side-effect of CodeSearch being in preview mode, or is the expectation that each search will only show the first 100 results?

As someone who has been using GitHub search for total counts, I am extremely interested to know if codesearch will be usable for determining total occurrences of terms across (unique) repositories. The 5 page limit with no indication of total counts makes it extremely difficult to use Code Search in its current form.

The first screenshot below looks way more useful to me than the 2nd: 

old search:

![image](https://user-images.githubusercontent.com/686/148660561-1d5db499-62e8-4a8c-9441-c114dd415951.png)

new search:

![image](https://user-images.githubusercontent.com/686/148660557-176f0b24-2f01-4aa2-9d8c-dd2f5a2e0ab4.png)

**Comment by anotherlab** [comment]
We have hundreds of repos that are on .NET 5. As part of our migration to .NET 6, it would be useful to be able to query for .csproj files that are .NET 5 and see all of them so that we can plan on our migrations.

**Comment by colinwm** [comment, marked as answer]
We are working on an API and intend to allow exhaustive search via the API without the 100 result limit.

As for estimating result set size, we've heard this feedback before and are working on it!

**Reply by 0x2b3bfa0** [reply]
It would be awesome to have this result count estimation. From that to the code homologous of [Google Ngram](https://books.google.com/ngrams) there is just a small step.

**Reply by iemejia** [reply]
@colinwm Is there already access to the REST API or a client implementation to access cs.github.com programmatically?

**Reply by colinwm** [reply]
Not yet, sorry! We absolutely will be releasing one though, stay tuned

**Reply by jpivarski** [reply]
I'm looking forward to this, too. I described my use-case in https://github.com/community/community/discussions/9991#discussioncomment-3863578 and today just tried to use it and noticed the limit of 100 results.

If it helps, I and maybe other users like me don't need all data about each result; a list of unique repositories would be enough for me, and other users might be content with a list of matching file names (and maybe also number of matches in each file).

**Reply by jpivarski** [reply]
It looks like the new code search is now the default (congratulations!), but now [this default search has become limited to 100 results](https://github.com/search?q=%22import+awkward%22+language%3APython&type=code&l=Python); it used to be 1000.

The REST API also seems to be limited to 100 results (or 104 in my case?). Is it using the new search? I fortunately have a saved copy of old results (813 of the estimated 4500 results), but I'd like to update it, and ideally get the repository names of all 4500 search hits.

Is there an expected date on when the search API will be able to do that? I'm writing an abstract for [a conference in May 2023](https://www.jlab.org/conference/CHEP2023) about how we can use GitHub tools to know how software packages are being used by our community and I want to know if I'll be able to update my results by then. Thanks!

**Reply by jpivarski** [reply]
Hi! I just want to check in on this. It has been a few months and the web search results are still limited to 100. The GitHub API's search (without `type=codelegacy`) stops after 10 pages and doesn't give quite 100 results per page.

Is there a plan to loosen these limits at some point? I'd like to know if this is how it's going to be indefinitely or if we're still in a temporary state.

Thanks!

**Reply by colinwm** [reply]
Hey @jpivarski, we still definitely intend to release an API for the new code search. Lately we've been focused on bringing the new code search to everyone in github.com.

However - we did implement result set counting, so you can get a pretty accurate count of how many results match a certain query. Does that help solve your problem re: usage stats? Or do you actually need to look at every single result?

**Reply by jpivarski** [reply]
The `"total_count"` field counts a lot more than the results I see.

For instance, I just did this query: `https://api.github.com/search/code?l=Python&q=%22import+awkward%22&per_page=100`, which had a `"total_count"` of 6458 but arrived in 3 pages of 4, 80, and 34, respectively.

As a test, I swapped `awkward` for `pandas`, which had a `"total_count"` of 6771317 but arrived in 10 pages of 58, 37, 44, 61, 78, 61, 78, 87, 40, and 92, respectively (total of 636). With the old search, it would have cut off after 1000 or 5000. (I don't remember anymore!)

For my purposes, I don't need the full repo records, I just need a list of repo names. If I wanted the repo records, I could do that in a separate query, but actually I want to clone the repos. (I won't be doing that for a popular search result like Pandas, of course.) The thing that I want to do with usage stats is to find out which functions in the library people are using, and I can spend some time cloning repos at a limited rate, then analyze them all offline. This is for finding out which parts of the library people are using, to direct more efforts in the directions that matter.

**Reply by seifeslimene** [reply]
> We are working on an API and intend to allow exhaustive search via the API without the 100 result limit.
> 
> As for estimating result set size, we've heard this feedback before and are working on it!

@colinwm You still working on it? can you give me an estimation for when this could be done?

**Reply by colinwm** [reply]
We intend to release the new API sometime this summer

**Reply by enriqueg** [reply]
Add to this that is impossible to search by index date and you have completely destroyed all the search feature. Congratulations! 👎

**Reply by Hua-Space** [reply]
> Add to this that is impossible to search by index date and you have completely destroyed all the search feature. Congratulations! 👎

Agree

**Reply by ldragone** [reply]
@colinwm is there any news about the new search API? 
Currently the code-search API is completely useless with these hard limits.

**Reply by arzedlab** [reply]
7 days left until the end of summer 😐

**Reply by Sh4d0wHunt3rX** [reply]
Any update on this?

**Reply by anthonyalayo** [reply]
Any update?

**Reply by falcucci** [reply]
they downgrade the search and we need to live like that now.

just give the option to rollback it at least. 

i am shocked

**Reply by beocaca** [reply]
Any update please...?

**Reply by StashOfCode** [reply]
@colinwm It's been two years lol "We are working..." my ass  XD

**Reply by cobavcova** [reply]
any update collin? please give back old code search

**Reply by NagaTheDon** [reply]
Any update on this?

**Reply by knhkse** [reply]
언제 업데이트함? 이러다 github 프리미엄 만들어서 돈 내야 전체 검색 결과 보게 해주는 기능 만들겠네

**Reply by Kaiser-Yang** [reply]
Three years has been, still working...

**Reply by Garador** [reply]
Ah, yes,the good ol' "We're working on this".

**Reply by kaiser101** [reply]
Wow it takes 3 years to develop an API to show the comprehensive set of results? Did you outsource building this API to Builder.ai?

**Comment by adonovan** [comment]
I understand the need for limits, but 100 seems very small. Large-scale refactoring and clean-up operations within an organization often touch hundreds of files or call sites. For larger result sets I'd be happy with just a list of affected files, or even repositories, so that I can use the search results as roots and do the rest in my scripts. Increasing this limit by a factor of 10 would be a welcome step. 

Regardless of the limit, it would also be very useful to be able to download the result set in a simple grep-like format ("file:line: text"). A dense result set would save the need to scrape filenames out of 5 or more HTML pages of results. I know that a 'gh search' tool is coming and will provide this feature, but it would be useful in the web interface too.

Furthermore, showing complete results within an organization's scope seems like an important feature for clean-up tasks, where you need to know that you've found all matches.

**Reply by stephdotnet** [reply]
Absolute yes for private repos in organization. This is my use case and wanted to use GH search in this scenario. Didn't since of this limitation 🤷

**Comment by catkins** [comment]
I'm running into this as well. We're trying to build some statistics across our org on adoption of a few specific internal libraries, but the limit of 100 results is meaning that I can't easily find my answers.

I'd be happy enough with the ability to paginate through all of them with a cursor and a lower rate limit as long as I could retrieve _all_ of the results.

**Comment by seifeslimene** [comment]
So is there any solution to the 100 results limit?

**Reply by Ponjimon** [reply]
The only solution right now is to narrow down your search. Unfortunately.

**Reply by falcucci** [reply]
they did rollout the new search and don't even offer a fallback solution.

that's really bad.

**Comment by falcucci** [comment]
unbelievable

**Comment by rhanton** [comment]
Ugh, this is crazy.

**Comment by adonovan** [comment]
> unbelievable

> Ugh, this is crazy.

Regardless of how much frustration you might feel, these are not constructive ways to vent it (and if you made these comments on my issue tracker, I would feel less, not more, inclined to help you). Bear in mind that software maintainers always have dozens of things competing for their attention. If this issue is important to you, argue the case.

**Comment by falcucci** [comment]
> Regardless of how much frustration you might feel, these are not constructive ways to vent it (and if you made these comments on my issue tracker, I would feel less, not more, inclined to help you). Bear in mind that software maintainers always have dozens of things competing for their attention. If this issue is important to you, argue the case.

seems that you didn't get that this has nothing with some specific developer/field. They are still here trying to help creating issues and using the platform.

It's just people tired about the mistakes the company did, messing one of the most used feature by them.

It's just a genuine expression about what happened here and how github is handling it.

The issue itself has enough detail about what they miss.

Unbelievable.

**Comment by StashOfCode** [comment]
It's just Microshit trying to prevent LLMs other than their chatgpt getting trained on open source code. Github is done.

**Comment by theharshpat** [comment]
hey what's up, any update on this?

**Comment by sgf** [comment]
I really can't understand Microsoft's behavior. You don't reduce costs through technology, but rely on castration of functions and degrading the experience. This is really puzzling.

**Comment by felix-d1strict** [comment]
@colinwm Sorry to ping you, but do you have any update on this?

**Comment by llakala** [comment]
Someone said to argue my case for this issue above, so here goes. 

Github search is really useful for finding real-world results that I can use as references for my code. Regardless of some of its flaws, it's an extraordinary tool. But its power is limited by the fact that it's unable to show more than 100 results. I wouldn't even complain that much if the limit was increased a bit for now as a temporary solution (I'd be happy with 250). I know that API development takes time, and that there are other priorities. But this is affecting a lot of users and would improve the experience of a lot of people. 

The solution doesn't need to be perfect immediately, but seeing any sort of improvement to this, even a minor one, would be much appreciated.

**Reply by theharshpat** [reply]
They don't want AI companies scrapping the data easily. But 100 is just way too less, it should be increased to 250 results at the very least.

**Reply by jpivarski** [reply]
The reason can't be to avoid scraping for AI, since all of the data is also available at https://www.gharchive.org/ (complete public metadata logs) and by git-cloning the repos directly (complete public git data). Both of these are in a more bulk-digestable form—good for AI training—than the search tool. The search tool is just for humans.

By the way, for my use-case of finding all users of my library, I now scrape the Insights/Dependency Graph (only available as HTML, as far as I can tell), then git-clone all of those repos and do an offline text search. GitHub is (thankfully!) very generous in how quickly it can deliver git data.

But this only underscores the point that the search tool is for humans. I wouldn't want to git-clone a repo just to search it for a keyword.

**Comment by sgf** [comment]
I hold the opposite view. I can only accept progress and cannot accept regression.

This should be obvious, I don't know why regression occurs. Is the existence of developers a reverse promotion effect? ​​If the developer is not capable enough, it is recommended to replace him. Don't stand there digging a hole without doing anything.

I don't know what problems GitHub has encountered since it was acquired by Microsoft. But it is obvious that many functions such as dashboard-feed have been reverse optimized. I seriously doubt whether Microsoft Group's management ability is insufficient.

Or maybe Microsoft is simply not good at creating and operating a large social network. Looking at Microsoft's history, I'm not making this up, but there are clues to follow.

Countless feedback from the community was created, but GitHub insisted on its own way, and these reverse optimizations had no precedent for improvement.

**Comment by theharshpat** [comment]
Please make it at least 250 results

**Comment by NagaTheDon** [comment]
Any update on this?? Why are we only allowed to see only 100 results? What happened to the old search?

**Comment by tbadalov** [comment]
Can't see beyond 5 pages of results. This is too small for the number of files actually found for a query. Hopefully this will get fixed soon.

**Comment by Floccinaucinihilipilification11** [comment]
For the past couple of weeks I suddenly started getting up to 100 pages of results (yes, up to 1000 results). I assume this is true for everyone?

**Reply by jpivarski** [reply]
Still just 5 pages (100 results) for me.

**Reply by Xaypanya** [reply]
Still 5 pages here

**Reply by Floccinaucinihilipilification11** [reply]
I get 100 pages with 10 results per page. Maybe it is slowly rolling out.

**Comment by radiomix** [comment]
I stumbled across a similar issue: GHA UI search window only finds/marks the first `100` search strings found! The UI is useless to count any thing inside logs that exists more often. 

I created a public repo to verify:  https://github.com/radiomix/gha-search-in-logs/actions/workflows/generate-search-string.yaml
You could just trigger the workflow and search in side the UI 

![image](https://github.com/user-attachments/assets/f6ab72fc-ec7e-45e0-97cd-85a704421320)

**Comment by Reikyo** [comment]
Just highlighting another [related issue](https://github.com/github/rest-api-description/issues/2956) here, which is that the browser and API searches over code (at least) don't necessarily return the same counts, and therefore possibly not the same results either. In my case, the browser search shows higher total counts than the API search, which is the same as the linked issue above. This is true even exploring use of `in:file`, `in:path` and `in:file,path` as documented [here](https://docs.github.com/en/search-github/searching-on-github/searching-code#search-by-the-file-contents-or-file-path). So even if the browser search is not capped to 100 results, will it return the correct and complete set of results? After trying the browser search and finding the issue of this thread, I then explored the API option, which is uncapped (EDIT: nope, [it is capped at 1,000 results](https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28)), but I found out about the count discrepancy instead. So I've simply switched one problem for another. I'm currently interested in counts rather than the exact results, but I can't trust either browser or API values if they don't agree with each other. It looks like there's some overall consolidation work that needs to be done across all GitHub search solutions.

**Reply by jpivarski** [reply]
I wonder if a third party could index GitHub for them. It's smaller than the internet as a whole and users of a GitHub-specific search would not be looking for fuzzy queries, and they might even be interested in specialized, historical searches ("which repos used to import library XYZ but dropped it?"). If the indexing is based on git, rather than GitHub-specific APIs, then maybe it could be applied to GitLab, BitBucket, sourcehut, etc. as well...

**Reply by Reikyo** [reply]
Yes a more general use solution would be welcome, regardless of who makes it. I find it a bit surprising that the overarching theme of this whole thread isn't somehow reflected in the [GitHub Public Roadmap](https://github.com/orgs/github/projects/4247/views/1), as far as I can see. Thorough and reliable search should be standard in 2025.

**Comment by 0xapp123** [comment]
Any updates?
I think, in the past, the github was shown over 1000 codes but from when did you set the limit?

**Reply by kaiser101** [reply]
Ever since they released the new version of search.

**Reply by 0xapp123** [reply]
Maybe 2~3 yrs ago, they only show 100 codes..

**Reply by kaiser101** [reply]
With the old search, they used to show all results. What Microsoft is doing is in return for free hosting of your code, they steal all your IP. Clever plan to buy Github. Only smart thing they did (or are capable of).
